### PR TITLE
Refactor contract deploy script

### PR DESCRIPTION
**Pull request by LEAFERx**
# Goals of PR

Core changes:

- Refactor contract deploy script to avoid using openzeppelin upgrades plugin to deploy proxy&impl directly

### CODE REVIEW

File: contracts/deploy/utils/index.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    );
+    const { doUpgrade } = await inquirer.prompt([
+      {
+        type: "confirm",
+        name: "doUpgrade",
+        message: "Do you want to upgrade the proxy?",
+        default: false,
+      },
+    ]);
+    if (!doUpgrade) {
```
Comment (destroyersrt) [inline contracts/deploy/utils/index.ts:44]: nit `Aborting upgradation` looks like a better message.

File: contracts/hardhat.config.ts [hunk 2]
```diff
@@ -71,19 +100,28 @@ const config: HardhatUserConfig = {
...
+      default: SEQUENCER_ADDRESS,
+      hardhat: 0,
+      localhost: 0,
+      specularLocalDev: 0,
     },
     validator: 1,
-    deployer: 2,
+    deployer: {
+      default: DEPLOYER_ADDRESS,
+      hardhat: 2,
```
[the remark this answers is not in the document]
Comment (LEAFERx) [reply contracts/hardhat.config.ts:111]: We only use these accounts in local (i.e. hardhat/localhost). if we do default of 2, we need to list every other networks here

File: contracts/hardhat.config.ts [hunk 3]
```diff
@@ -18,18 +19,41 @@ const ALCHEMY_KEY = process.env.ALCHEMY_KEY || "";
...
-  if (network === "goerli" || network === "chiado") {
-    return SEQUENCER_PRIVATE_KEY
-      ? [`0x${SEQUENCER_PRIVATE_KEY}`]
+  if (
+    network === "goerli" ||
+    network === "chiado" ||
+    network === "specularDev"
+  ) {
+    return DEPLOYER_PRIVATE_KEY
+      ? [`0x${DEPLOYER_PRIVATE_KEY}`, `0x${SEQUENCER_PRIVATE_KEY}`]
```
[the remark this answers is not in the document]
Comment (LEAFERx) [reply contracts/hardhat.config.ts:56]: the index is not relevant here

File: contracts/hardhat.config.ts [hunk 4]
```diff
@@ -18,18 +19,41 @@ const ALCHEMY_KEY = process.env.ALCHEMY_KEY || "";
 const SEQUENCER_PRIVATE_KEY = process.env.SEQUENCER_PRIVATE_KEY;
 const ETHERSCAN_API_KEY = process.env.ETHERSCAN_API_KEY || "";
 const SEQUENCER_ADDRESS = process.env.SEQUENCER_ADDRESS ?? wallet.address;
+const DEPLOYER_PRIVATE_KEY = process.env.DEPLOYER_PRIVATE_KEY;
+const DEPLOYER_ADDRESS = process.env.DEPLOYER_ADDRESS ?? wallet.address;
 
 function createConfig(network: string) {
-  return {
+  const live = network !== "localhost" && network !== "specularLocalDev";
```
[the remark this answers is not in the document]
Comment (LEAFERx) [reply contracts/hardhat.config.ts:26]: hardhat network config is not created by this func

File: contracts/hardhat.config.ts [hunk 5]
```diff
@@ -71,19 +100,28 @@ const config: HardhatUserConfig = {
...
+    deployer: {
+      default: DEPLOYER_ADDRESS,
+      hardhat: 2,
+      localhost: 2,
+      specularLocalDev: 2,
+    },
   },
   networks: {
     mainnet: createConfig("mainnet"),
     sepolia: createConfig("sepolia"),
```
[the remark this answers is not in the document]
Comment (LEAFERx) [reply contracts/hardhat.config.ts:118]: we might want to later

File: contracts/hardhat.config.ts [hunk 6]
```diff
@@ -46,14 +70,19 @@ function getNetworkURL(network: string) {
...
     return "https://rpc.gnosischain.com/";
   } else if (network === "chiado") {
-    return "https://rpc.chiadochain.net";
+    return "https://rpc.chiado.gnosis.gateway.fm";
+  } else if (network === "specularLocalDev") {
+    return "http://localhost:4011";
+  } else if (network === "specularDev") {
+    return "https://devnet.specular.network";
   }
+  return "http://localhost:8545";
```
[the remark this answers is not in the document]
Comment (LEAFERx) [reply contracts/hardhat.config.ts:79]: hardhat is the local network but local network might not always be hardhat

File: contracts/deploy/utils/index.ts (hunk 1 above)
Comment (LEAFERx) [reply contracts/deploy/utils/index.ts:44]: the entire deployment process is aborted

File: contracts/hardhat.config.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (destroyersrt) [reply contracts/hardhat.config.ts:79]: but the localhost is defined as hardhat when calling createConfig.

File: contracts/hardhat.config.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (destroyersrt) [reply contracts/hardhat.config.ts:111]: with this we will have to change the default network each time, right?

File: contracts/hardhat.config.ts [hunk 7]
```diff
@@ -18,18 +19,41 @@ const ALCHEMY_KEY = process.env.ALCHEMY_KEY || "";
 const SEQUENCER_PRIVATE_KEY = process.env.SEQUENCER_PRIVATE_KEY;
 const ETHERSCAN_API_KEY = process.env.ETHERSCAN_API_KEY || "";
 const SEQUENCER_ADDRESS = process.env.SEQUENCER_ADDRESS ?? wallet.address;
+const DEPLOYER_PRIVATE_KEY = process.env.DEPLOYER_PRIVATE_KEY;
+const DEPLOYER_ADDRESS = process.env.DEPLOYER_ADDRESS ?? wallet.address;
 
 function createConfig(network: string) {
```
Comment (ujvl) [inline contracts/hardhat.config.ts:25]: need to add some documentation here about what these different networks mean, it's a bit confusing

File: contracts/hardhat.config.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ujvl) [reply contracts/hardhat.config.ts:56]: yes should probably check `SEQUENCER_PRIVATE_KEY && DEPLOYER_PRIVATE_KEY` right?

File: contracts/hardhat.config.ts [hunk 8]
```diff
@@ -18,18 +19,41 @@ const ALCHEMY_KEY = process.env.ALCHEMY_KEY || "";
...
+  } else if (network === "specularDev") {
+    config.companionNetworks = { l1: "chiado" };
+    config.chainId = 93481;
+  } else if (network === "chiado") {
+    config.chainId = 10200;
+  }
+  return config;
 }
 
 function getNetworkAccounts(network: string) {
```
Comment (ujvl) [inline contracts/hardhat.config.ts:49]: i'd also add some documentation here about these different cases

**Review by ujvl — APPROVED** [review]
some more documentation & see my other comments on the other PR as well but otherwise lgtm

**Pull request merged**